Reject corrupt stream RDB payloads with a NACK shared across consumers.

A crafted stream RESTORE payload can make two consumers reference the same
pending-entries-list NACK; acknowledging it from one consumer frees the NACK
while the other still points at it (use-after-free). Refuse to load a
consumer PEL entry whose NACK already has a consumer assigned.

Tracker: none (Redis 6.2.23 security release)
Upstream patch: https://github.com/redis/redis/commit/41a958720e64e03576dd652d224aa46d22c096c3
Adaptation: src/rdb.c -> src/rdb.cpp; the existing "duplicated consumer PEL entry" test (introduced here by the CVE-2026-25243 backport) now expects the new guard's log line, as upstream changed it in the same commit.
diff --git a/src/rdb.cpp b/src/rdb.cpp
--- a/src/rdb.cpp
+++ b/src/rdb.cpp
@@ -2482,6 +2482,15 @@
                         return NULL;
                     }
 
+                    /* If the NACK already has a consumer assigned, the
+                     * payload is corrupt: each global PEL entry must be
+                     * claimed by exactly one consumer. */
+                    if (nack->consumer != NULL) {
+                        rdbReportCorruptRDB("Stream consumer PEL entry already has a consumer assigned");
+                        decrRefCount(o);
+                        return NULL;
+                    }
+
                     /* Set the NACK consumer, that was left to NULL when
                      * loading the global PEL. Then set the same shared
                      * NACK structure also in the consumer-specific PEL. */
diff --git a/tests/integration/corrupt-dump.tcl b/tests/integration/corrupt-dump.tcl
--- a/tests/integration/corrupt-dump.tcl
+++ b/tests/integration/corrupt-dump.tcl
@@ -89,7 +89,7 @@
             r restore k 0 "\x0f\x01\x10\x00\x00\x01\x9b\x73\x06\x3b\xed\x00\x00\x00\x00\x00\x00\x00\x00\x28\x28\x00\x00\x00\x0f\x00\x02\x01\x00\x01\x01\x01\x81\x66\x02\x00\x01\x02\x01\x00\x01\x00\x01\x81\x76\x02\x04\x01\x02\x01\x00\x01\x01\x01\x81\x76\x02\x04\x01\xff\x02\x81\x00\x00\x01\x9b\x73\x06\x3b\xed\x01\x01\x01\x67\x81\x00\x00\x01\x9b\x73\x06\x3b\xed\x01\x02\x00\x00\x01\x9b\x73\x06\x3b\xed\x00\x00\x00\x00\x00\x00\x00\x00\xee\x3b\x06\x73\x9b\x01\x00\x00\x01\x00\x00\x01\x9b\x73\x06\x3b\xed\x00\x00\x00\x00\x00\x00\x00\x01\xee\x3b\x06\x73\x9b\x01\x00\x00\x01\x01\x01\x63\xee\x3b\x06\x73\x9b\x01\x00\x00\x02\x00\x00\x01\x9b\x73\x06\x3b\xed\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x9b\x73\x06\x3b\xed\x00\x00\x00\x00\x00\x00\x00\x00\x09\x00\x59\xa9\xd3\x45\x4d\xc3\x9a\x5b" REPLACE
         } err
         assert_match "*Bad data format*" $err
-        verify_log_message 0 "*Duplicated consumer PEL entry*" 0
+        verify_log_message 0 "*Stream consumer PEL entry already has a consumer assigned*" 0
     }
 }
 
@@ -778,6 +778,22 @@
         r debug set-skip-checksum-validation 1
         r config set sanitize-dump-payload yes
         catch { r restore _stream 0 "\x0F\x03\x10\x00\x00\x01\x7B\x60\x5A\x23\x79\x00\x00\x00\x00\x00\x00\x00\x00\xC3\x40\x4F\x40\x5C\x18\x5C\x00\x00\x00\x24\x00\x05\x01\x00\x01\x02\x01\x84\x69\x74\x65\x6D\x05\x85\x76\x61\x6C\x75\x65\x06\x40\x10\x00\x00\x20\x01\x00\x01\x20\x03\x00\x05\x20\x1C\x40\x09\x05\x01\x01\x82\x5F\x31\x03\x80\x0D\x00\x02\x20\x0D\x00\x02\xA0\x19\x00\x03\x20\x0B\x02\x82\x5F\x33\xA0\x19\x00\x04\x20\x0D\x00\x04\x20\x19\x00\xFF\x10\x00\x00\x01\x7B\x60\x5A\x23\x79\x00\x00\x00\x00\x00\x00\x00\x05\xC3\x40\x51\x40\x5E\x18\x5E\x00\x00\x00\x24\x00\x05\x01\x00\x01\x02\x01\x84\x69\x74\x65\x6D\x05\x85\x76\x61\x6C\x75\x65\x06\x40\x10\x00\x00\x20\x01\x06\x01\x01\x82\x5F\x35\x03\x05\x20\x1E\x40\x0B\x03\x01\x01\x06\x01\x80\x0B\x00\x02\x20\x0B\x02\x82\x5F\x37\xA0\x19\x00\x03\x20\x0D\x00\x08\xA0\x19\x00\x04\x20\x0B\x02\x82\x5F\x39\x20\x19\x00\xFF\x10\x00\x00\x01\x7B\x60\x5A\x23\x79\x00\x00\x00\x00\x00\x00\x00\x00\xC3\x3B\x40\x49\x18\x49\x00\x00\x00\x15\x00\x02\x01\x00\x01\x02\x01\x84\x69\x74\x65\x6D\x05\x85\x76\x61\x6C\x75\x65\x06\x40\x10\x00\x00\x20\x01\x40\x00\x00\x05\x20\x07\x40\x09\xC0\x22\x09\x01\x01\x86\x75\x6E\x69\x71\x75\x65\x07\xA0\x2C\x02\x08\x01\xFF\x0C\x81\x00\x00\x01\x7B\x60\x5A\x23\x7A\x01\x00\x09\x00\x9C\x8F\x1E\xBF\x2E\x05\x59\x09" replace } err
+        assert_match "*Bad data format*" $err
+        r ping
+    }
+}
+
+test {corrupt payload: stream with NACK shared between two consumers} {
+    start_server [list overrides [list loglevel verbose use-exit-on-panic yes crash-memcheck-enabled no]] {
+        r debug set-skip-checksum-validation 1
+        # Payload: stream with entry 1-0, one consumer group (mygroup),
+        # two consumers whose PELs both reference 1-0 (shared NACK).
+        # XACK on one consumer frees the NACK, leaving a dangling
+        # pointer in the other consumer's PEL (use-after-free).
+        catch {r RESTORE mystream 0 "\x0f\x01\x10\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x25\x25\x00\x00\x00\x0a\x00\x01\x01\x00\x01\x01\x01\x85\x66\x69\x65\x6c\x64\x06\x00\x01\x02\x01\x00\x01\x00\x01\x85\x76\x61\x6c\x75\x65\x06\x04\x01\xff\x01\x01\x00\x01\x07\x6d\x79\x67\x72\x6f\x75\x70\x01\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x4a\x79\x89\x8f\x9f\x01\x00\x00\x01\x02\x09\x63\x6f\x6e\x73\x75\x6d\x65\x72\x41\x4a\x79\x89\x8f\x9f\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x09\x63\x6f\x6e\x73\x75\x6d\x65\x72\x42\x54\x79\x89\x8f\x9f\x01\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x09\x00\xf6\xea\x7b\x6c\xb8\xb8\xd8\xb4"} err
+        catch {r XACK mystream mygroup 1-0} _
+        catch {r XREADGROUP GROUP mygroup consumerA COUNT 10 STREAMS mystream 0} _
+        catch {r DEL mystream} _
         assert_match "*Bad data format*" $err
         r ping
     }
